server/taskAnomaliesDetection.go

```diff
@@ -0,0 +1,187 @@
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	ea := models.EmailedAnomaly{
+		Date:         date,
+		Recipient:    user.Email,
+		AwsAccountID: awsAccount.Id,
+		Product:      product,
+	}
+	var subject = fmt.Sprintf("TrackIt detected an abnormal peak in your %s cost on your account %s!", product, awsAccount.Pretty)
+	var body = fmt.Sprintf(`
+TrackIt detected an abnormal peak in your spendings!
+
+		Which account? %s
+		Which product? %s
+		How much? %f
+		How important? %.1f
+		When? %s</p>
+See more on our website http://re.trackit.io/
+
+this email is originally intended for %s.
+`, awsAccount.Pretty, product, an.Cost, an.Cost / an.UpperBand * 100, date.String(), ea.Recipient)
+	var recipient = "team@trackit.io" // replace by ea.Recipient.
+	if err := mail.SendMail(recipient, subject, body, ctx); err != nil {
+		logger.Error("Error when sending mail", err)
+		return err
+	} else if err := ea.Insert(db.Db); err != nil {
+		logger.Error("Error when updating db with the sent email", err)
+		return err
+	} else {
+		logger.Info("Email sent", map[string]interface{}{
+			"aws_account": map[string]interface{}{
+				"id":       awsAccount.Id,
+				"role_arn": awsAccount.RoleArn,
+				"pretty":   awsAccount.Pretty,
+				"user_id":  awsAccount.UserId,
+			},
+			"anomaly": map[string]interface{}{
+				"date": an.Date,
+				"cost": an.Cost,
+			},
+			"mail": map[string]interface{}{
+				"recipient": recipient,
+				"subject":   subject,
+				"body":      body,
+			},
+		})
+	}
+	return nil
+}
+
+// processAnomaliesByProduct checks if the anomaly has already been sent.
+// If not, it sends it.
+func processAnomaliesByProduct(user users.User, awsAccount aws.AwsAccount, product string, ans []anomalies.CostAnomaly, tx *sql.Tx, ctx context.Context) error {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	for _, an := range ans {
+		if !an.Abnormal {
+		} else if date, err := time.Parse("2006-01-02T15:04:05.000Z", an.Date); err != nil {
+			logger.Error("Error when parsing date", err)
+			return err
+		} else if alreadyEmailed, err := models.IsAnomalyAlreadyEmailed(tx, awsAccount.Id, product, date); err != nil {
+			logger.Error("Error when checking db for sent email", err)
+			return err
+		} else if alreadyEmailed {
+		} else if err := sendAnomalyEmail(user, awsAccount, product, an, date, ctx); err != nil {
+			return nil
+		}
+	}
+	return nil
+}
+
+// requestAnomalies requests anomalies from GetAnomaliesData and calls processAnomaliesByProduct.
+func requestAnomalies(user users.User, awsAccount aws.AwsAccount, tx *sql.Tx, ctx context.Context) error {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	identity, err := awsAccount.GetAwsAccountIdentity()
+	if err != nil {
+		logger.Error("Error when getting identity from AwsAccount", map[string]interface{}{
+			"awsAccount": awsAccount,
+			"error":      err,
+		})
+		return err
+	}
+	params := anomalies.AnomalyEsQueryParams{
+		DateBegin:   time.Now().Add(-1 * 24 * 7 * time.Hour),
```

Comment:
Keep in mind that the tasks are not started at a fixed time, we will miss data if we request only the last week. We should start on the time of the last anomaly detection

Reply:
The task has to be launched at least every week so we won't miss anything.
We could miss an anomaly if we launch it every 2 weeks for example.
The goal is to launch it every 12h so the anomalies are quickly detected and emailed.

Reply:
ok!